pkg/liqo-controller-manager/namespace-controller/namespace_controller.go

```diff
@@ -0,0 +1,270 @@
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"strings"
+)
+
+type NamespaceReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+const (
+	mappingLabel          = "mapping.liqo.io"
+	offloadingLabel       = "offloading.liqo.io"
+	offloadingPrefixLabel = "offloading.liqo.io/"
+	namespaceFinalizer    = "namespace.liqo.io/finalizer"
+)
+
+func (r *NamespaceReconciler) removeRemoteNamespace(localName string, nm *mapsv1alpha1.NamespaceMap) error {
+
+	if _, ok := nm.Status.NattingTable[localName]; ok {
+
+		delete(nm.Status.NattingTable, localName)
+		if err := r.Update(context.TODO(), nm); err != nil {
+			klog.Error(err, " --> Unable to update NamespaceMap")
+			return err
+		}
+		klog.Info(" Entries deleted correctly")
+
+	}
+
+	return nil
+}
+
+func (r *NamespaceReconciler) removeRemoteNamespaces(localName string, nms map[string]*mapsv1alpha1.NamespaceMap) error {
+
+	for _, nm := range nms {
+		if err := r.removeRemoteNamespace(localName, nm); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *NamespaceReconciler) createRemoteNamespace(n *corev1.Namespace, remoteName string, nm *mapsv1alpha1.NamespaceMap) error {
+
+	if nm.Status.NattingTable == nil {
+		nm.Status.NattingTable = map[string]string{}
+	}
+
+	if oldValue, ok := nm.Status.NattingTable[n.GetName()]; ok {
+		if oldValue != remoteName {
+			n.GetLabels()[mappingLabel] = oldValue
+			if err := r.Update(context.TODO(), n); err != nil {
+				klog.Error(err, " --> Unable to update mapping label")
+				return err
+			}
+		}
+	} else {
+		nm.Status.NattingTable[n.GetName()] = remoteName
+		if err := r.Patch(context.TODO(), nm, client.Merge); err != nil {
+			klog.Error(err, " --> Unable to add entries in NamespaceMap")
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *NamespaceReconciler) createRemoteNamespaces(n *corev1.Namespace, remoteName string, nms *mapsv1alpha1.NamespaceMapList) error {
+
+	for _, nm := range nms.Items {
+		if err := r.createRemoteNamespace(n, remoteName, &nm); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *NamespaceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+
+	namespace := &corev1.Namespace{}
+	if err := r.Get(context.TODO(), req.NamespacedName, namespace); err != nil {
+		klog.Error(err, " --> Unable to get namespace")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	namespaceMaps := &mapsv1alpha1.NamespaceMapList{}
+	if err := r.List(context.TODO(), namespaceMaps); err != nil {
+		klog.Error(err, " --> Unable to List NamespaceMaps")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if len(namespaceMaps.Items) == 0 {
+		klog.Info(" No namespaceMaps at the moment")
+		return ctrl.Result{}, nil
+	}
+
+	removeMappings := make(map[string]*mapsv1alpha1.NamespaceMap)
```

Comment:
Only for efficiency reasons, even if probably there won't be many `NamespaceMaps` in the cluster... Do you think is better to use directly `NamespaceMap` instead of `*NamespaceMap`

Reply:
A map in Go is a reference type. This means that every time you pass a map to a function, you pass the reference to that map instead of its copy.